src/interchange/src/protobuf/decode.rs

```diff
@@ -233,50 +237,101 @@ fn json_nested_from_serde_value(
         }
         SerdeValue::Map(m) => {
-            let mut kvs = m.iter().collect::<Vec<_>>();
-            kvs.sort_by(|(k1, _v1), (k2, _v2)| k1.cmp(k2));
-            kvs.dedup_by(|(k1, _v1), (k2, _v2)| k1 == k2);
-            return packer.push_dict_with(|packer| {
-                let nested_message_descriptor = f.field_type(descriptors);
-                for (k, v) in kvs {
-                    match k {
-                        SerdeValue::String(s) => {
+            let nested_message_descriptor = f.field_type(descriptors);
+            let pack_map = |packer: &mut Row,
+                            k: &SerdeValue,
+                            v: &SerdeValue,
+                            typ: &ColumnType,
+                            push_key: bool| {
+                match k {
+                    SerdeValue::String(s) => {
+                        if push_key {
                             packer.push(Datum::String(s.as_str()));
+                        }
 
-                            let nested_message_descriptor = match nested_message_descriptor {
-                                FieldType::Message(m) => m,
-                                _ => bail!("Nested message is the wrong type"),
-                            };
+                        let nested_message_descriptor = match nested_message_descriptor {
+                            FieldType::Message(m) => m,
+                            _ => bail!("Nested message is the wrong type"),
+                        };
 
-                            json_nested_from_serde_value(
-                                &v,
-                                packer,
-                                nested_message_descriptor
-                                    .field_by_name(s)
-                                    .expect("Expected this to work"),
-                                descriptors,
-                            )?;
-                        }
-                        _ => bail!("Unrecognized value while trying to parse a nested message"),
+                        json_nested_from_serde_value(
+                            &v,
+                            packer,
+                            nested_message_descriptor
+                                .field_by_name(s)
+                                .expect("nested message to exist"),
+                            descriptors,
+                            typ,
+                        )?;
                     }
+                    _ => bail!("Unrecognized value while trying to parse a nested message"),
                 }
                 Ok(())
-            });
+            };
+
+            match &column_type.scalar_type {
+                ScalarType::Map { value_type, .. } => {
+                    let mut kvs = m.iter().collect::<Vec<_>>();
+                    kvs.sort_by(|(k1, _v1), (k2, _v2)| k1.cmp(k2));
+                    kvs.dedup_by(|(k1, _v1), (k2, _v2)| k1 == k2);
```

Comment:
I'm a little confused why the dedup is necessary? I would expect that either protobuf records aren't allowed to have dups or that we should throw an error stating that we don't support duplicates if there are dups or something.